### PR TITLE
Fix inner type namespace issue

### PR DESCRIPTION
**Pull request by xiaoyu-yang-gh**
Starting from Avro 1.7, if a schema has namespace for a outer type, and inner record type does not have namespace, the printed avsc adds an empty string as namespace for the inner type.

We fix the SchemaBuilder to recursively populate the outer namespace to the inner types if the inner types do not have namespace. If a inner type already has namespace, the SchemaBuilder does not overwrite its namespace.

### CODE REVIEW

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java [hunk 1]
```diff
@@ -250,6 +260,49 @@ public Schema build() {
...
+                        .setElementType(schema.getElementType())
+                        .setNamespace(parentNamespace)
+                        .build();
+                }
+                break;
+            case MAP:
+                if (parentNamespace != null) {
+                    result = _adapter.newSchemaBuilder(null)
+                        .setType(Schema.Type.MAP)
+                        .setValueType(schema.getValueType())
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:287]: you might need to recurse into the value type.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java [hunk 2]
```diff
@@ -250,6 +260,49 @@ public Schema build() {
...
+                        .setNamespace(parentNamespace)
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:288]: what does namespace mean for a map schema?

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java [hunk 3]
```diff
@@ -250,6 +260,49 @@ public Schema build() {
...
+                        .build();
+                }
+                break;
+            case UNION:
+                if (parentNamespace != null) {
+                    result = _adapter.newSchemaBuilder(null)
+                        .setType(Schema.Type.UNION)
+                        .setUnionBranches(schema.getTypes())
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:296]: you will need to recurse into the union branches

File: helper/tests/helper-tests-common/src/main/resources/CCDV2.avsc [hunk 4]
```diff
@@ -0,0 +1,86 @@
+{
+  "type": "record",
```
Comment (radai-rosenblatt) [inline helper/tests/helper-tests-common/src/main/resources/CCDV2.avsc:2]: this schema is "too complex" and also lacks unions maps and lists for corner cases

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java [hunk 5]
```diff
@@ -250,6 +260,49 @@ public Schema build() {
...
+            case RECORD:
+                if (schema.getNamespace() == null && parentNamespace != null) {
+                    result = _adapter.newSchemaBuilder(schema).setNamespace(parentNamespace).build();
+                }
+                break;
+            case ARRAY:
+                if (parentNamespace != null) {
+                    result = _adapter.newSchemaBuilder(null)
+                        .setType(Schema.Type.ARRAY)
+                        .setElementType(schema.getElementType())
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:278]: you need to recurse into element type

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java (hunk 2 above)
Comment (xiaoyu-yang-gh) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:288]: For Map, union, array, the parentNamespace is passed in to indicate the closest ancestor that has the namespace, so when processing the map schema, the namespace will be passed along when processing its value schema.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java (hunk 1 above)
Comment (xiaoyu-yang-gh) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:287]: **recurse into the value type** happens in the build() method.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java (hunk 5 above)
Comment (xiaoyu-yang-gh) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:278]: in the build() method.

File: helper/tests/helper-tests-common/src/main/resources/CCDV2.avsc (hunk 4 above)
Comment (xiaoyu-yang-gh) [reply helper/tests/helper-tests-common/src/main/resources/CCDV2.avsc:2]: we have the union in the test. will check other cases.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java [hunk 6]
```diff
@@ -250,6 +291,26 @@ public Schema build() {
...
+    private Schema inheritNamespace(Schema schema) {
+        if (!_inheritNamespace) {
+            return schema;
+        }
+
+        SchemaBuilder schemaBuilder = _adapter.newSchemaBuilder(schema).setInheritNamespace(_inheritNamespace);
+
+        //NamedSchema inherits parent's namespace only if it does not have namespace
+        Schema.Type schemaType = schema.getType();
+        if (schemaType == Schema.Type.FIXED || schemaType == Schema.Type.ENUM || schemaType == Schema.Type.RECORD) {
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java:303]: we have a set of NAMED_TYPES somewhere you could use instead of this comparison chain

Comment (xiaoyu-yang-gh) [reply helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AbstractSchemaBuilder.java]: changed to use NAMED_TYPES.

**Review by radai-rosenblatt — APPROVED** [review]
